src/sms.rs

```diff
@@ -0,0 +1,201 @@
+    fn ascii_to_gsm7bit<const N: usize>(text: &str) -> Result<String<N>, Error> {
+        let mut encoded_message = String::new();
+
+        for c in text.chars() {
+            if c.is_ascii() {
+                let char_7bit = ASCII_TO_7BIT_TABLE[c as usize];
+                if char_7bit & STR_7BIT_ESCAPE_IND == 0 {
+                    encoded_message
+                        .push(char_7bit as char)
+                        .map_err(|_| Error::BufferTooSmall(None))?;
+                } else {
+                    encoded_message
+                        .push(STR_7BIT_ESCAPE_CODE as char)
+                        .map_err(|_| Error::BufferTooSmall(None))?;
+                    encoded_message
+                        .push((char_7bit & STR_7BIT_CODE_MASK) as char)
+                        .map_err(|_| Error::BufferTooSmall(None))?;
+                }
+            }
+        }
+
+        Ok(encoded_message)
+    }
+    // Pack a GSM 7 bit strings into 7 bites without 1 bit padding
+    // Reimplement from https://github.com/nrfconnect/sdk-nrf/blob/main/lib/sms/string_conversion.c#L294
+    fn pack_gsm7bit<const N: usize>(text: String<N>) -> Vec<u8, N> {
+        let mut src: usize = 0;
+        let mut dst: usize = 0;
+        let mut shift: usize = 0;
+        let len = text.len();
+        let mut bytes = text.into_bytes();
+
+        while src < len {
+            bytes[dst] = bytes[src] >> shift;
+            src += 1;
+            if src < len {
+                bytes[dst] |= bytes[src] << (7 - shift);
+                shift += 1;
+                if shift == 7 {
+                    shift = 0;
+                    src += 1;
+                }
+            }
+            dst += 1;
+        }
+        bytes.truncate(dst);
+        bytes
+    }
+    /// Sends the craftes message
+    /// `N` is need to provide internal buffer size for message and number encoding. Needs to be at least 2 * message.len() + 34
+    pub async fn send<const N: usize>(self) -> Result<(), Error> {
+        let encoded_number = Self::encode_number(self.number)?;
+
+        let encoded_message = Self::pack_gsm7bit(Self::ascii_to_gsm7bit::<N>(self.message)?);
+
+        let size = 2 + /* First header byte and TP-MR fields */
+		1 + /* Length of phone number */
+		1 + /* Phone number Type-of-Address byte */
+		encoded_number.len()/2 +
+		2 + /* TP-PID and TP-DCS fields */
+		1 + /* TP-UDL field */
+		encoded_message.len();
+
+        let mut at_cmgs: String<N> = String::new();
+        let mut encoded_number_len = encoded_number.len();
+        if self.number.trim_start_matches('+').len() % 2 != 0 {
+            encoded_number_len -= 1;
+        }
+        // Write the at command and begin with encoded number and it's lenght
+        write!(
+            &mut at_cmgs,
+            "AT+CMGS={}\r{:04X}{:04X}91{}",
+            size, 0x01, encoded_number_len, encoded_number
+        )
```

Comment:
Would it be possible to replace these writes with my `at-commands` crate? It would be nice to avoid this expensive formatting. If there's no other integer formatting in the project, this could easily add 2-3kb flash usage.

Reply:
I am looking at at-commands, looks pretty nice but I think it would need to be extended to do that. The PDU formatting of the message to send is a mess.
I think the simplest way would be to add something like `finish_with_data(at_terminator: &[u8], data: &u8, data_terminator: &[u8])` and leave the PDU string formatting outside `at-commands`.
That still leaves us with something that we need to format numbers to hex, that is used a lot in this. Another option would be to use ufmt https://crates.io/crates/ufmt with I was thinking about but I didn't want at first pull a extra crate for that.
What do you think about this?

Reply:
Ok, well if it isn't possible now, then I think you could leave it as it is right now. It's always possible to improve it in the future too